python-avd/pyavd/api/pool_manager/base_classes.py

```diff
@@ -0,0 +1,282 @@
+
+        file_data = load(self.pools_file.read_text(encoding="UTF-8"), Loader=CSafeLoader)
+        if not isinstance(file_data, dict) or self.pools_key not in file_data:
+            self._pools = {}
+            return
+
+        try:
+            self._pools = self._pools_from_list(file_data[self.pools_key])
+        except (TypeError, KeyError, ValueError) as e:
+            msg = f"An error occurred during parsing of '{self.pools_file}': {e.__class__.__name__}: {e}"
+            raise TypeError(msg) from e
+
+    def _pools_from_list(self, data: list) -> dict[T_PoolKeyType, Pool[T_PoolKeyType, T_AssignmentKeyType, T_ValueType]]:
+        """Returns pools from file data."""
+        if not isinstance(data, list):
+            msg = f"Invalid type '{type(data)}'. Expected a list."
+            raise TypeError(msg)
+        return {pool.pool_key: pool for item in data if (pool := self.pool_cls.from_dict(item, collection=self))}
+
+    def get_pool(self, pool_key: T_PoolKeyType) -> Pool[T_PoolKeyType, T_AssignmentKeyType, T_ValueType]:
+        """
+        Returns the pool for the given key if found in the collection.
+
+        Otherwise a new pool is inserted into the collection and returned.
+        """
+        if pool_key not in self._pools:
+            self._pools[pool_key] = self.pool_cls(collection=self, pool_key=pool_key, assignments={})
+            self.changed = True
+
+        return self._pools[pool_key]
+
+    def _remove_stale_assignments(self) -> None:
+        """
+        Remove stale assignments from all pools in the collection and remove empty pools afterwards.
+
+        Note: This method is called from the parent pool manager and should _not_ be called manually.
+        """
+        len_before = len(self._pools)
+        [pool._remove_stale_assignments() for pool in self._pools.values()]
+        self._pools = {pool_key: pool for pool_key, pool in self._pools.items() if pool.assignments}
+        self.changed = self.changed or len_before != len(self._pools)
+
+    def as_list(self) -> list[dict]:
+        """Returns a list of dicts representing the object."""
+        return natural_sort([pool.as_dict() for pool in self._pools.values()], sort_key="pool_key")
+
+    def save_updates(self, dumper_cls: type = CSafeDumper) -> bool:
+        """
+        Save data if anything changed. Returns a boolean telling if anything was changed.
+
+        Calls self._remove_stale_assignments first to ensure we save the cleaned data.
+
+        Data is sorted to ensure a consistent layout.
+        """
+        self._remove_stale_assignments()
+        if not self.changed:
+            return False
+
+        if not self.pools_file.exists():
+            # Try to create the dir and file.
+            self.pools_file.parent.mkdir(mode=0o775, parents=True, exist_ok=True)
+            self.pools_file.touch(mode=0o664)
+
+        self.pools_file.write_text(FILE_HEADER + dump({self.pools_key: self.as_list()}, Dumper=dumper_cls))
```

Comment:
need to handle permission error / OSError

Reply:
recasting with some extra information